README.md

```diff
@@ -48,33 +65,470 @@ cache read/white process, etc. Command handler can be sync and publish the
+      wait: { [weak self] in
+        // Wait for results to come.
+        self?.refreshControl?.beginRefreshing()
+      },
+      consume: { [weak self] books in
+        // We're lucky, there are some books to display.
+        self?.books = books
+        self?.refreshControl?.endRefreshing()
+        self?.tableView.reloadData()
+      },
+      rescue: { [weak self] error in
+        // Well, seems like something went wrong.
+        self?.refreshControl?.endRefreshing()
+        print(error)
+      }))
+  }
+
+  // ...
+}
+```
+
+**It's important** to dispose all reaction tokens when your `ReactionProducer`
+instance is about to be deallocated or reaction needs to be unsubscribed from
+events.
+
+```swift
+// Disposes all reaction tokens for the current `ReactionProducer`.
+disposeAll()
+
+// Disposes a specified reaction token.
+let token = react(to: BookListCommand.self, with: reaction)
+dispose(token: token)
+```
+
+## Extra
+
+### Action
+
+**Action** is a variation of command that handles itself. It's a possibility to
+simplify the code when command itself or business logic are super tiny. There
+is no need to register an action, it will be automatically added to the list of
+active command handlers on the fly, when it's executed as a command.
+
+```swift
+import Sugar
+
+struct WelcomeAction: Action {
+  typealias Output = String
+
+  let userId: String
+
+  func handle(command: WelcomeAction) throws -> Event<WelcomeAction> {
+    fetchUser(id: userId) { user in
+      self.publish(data: "Hello \(user.name)")
+    }
+    return Event.Progress
+  }
+}
+
+// Execute action
+
+struct WelcomeManager: CommandProducer {
+
+  func salute() {
+    execute(WelcomeAction(userId: 11))
+  }
+}
+```
+
+### Fact
+
+**Fact** works like notification with no async operations involved. It could
+be used when there is no need to have a handler and generate an output. Fact is
+already an output itself, so the only thing you want to do is notify all
+subscribers that something happened in the system, and they will react
+accordingly. In this sense it's closer to a type-safe alternative to
+`NSNotification`.
+
+```swift
+struct LoginFact: Fact {
+  let username: String
+}
+
+class ProfileController: UIViewController, ReactionProducer {
+
+  override func viewDidLoad() {
+    super.viewDidLoad()
+
+    // React
+    next { (fact: LoginFact) in
+      title = fact.username
+    }
+  }
+}
+
+struct AuthService: FactProducer {
+
+  func login() {
+    let fact = LoginFact(username: "John Doe")
+    // Publish
+    post(fact: fact)  
+  }
+}
+```
+
+### Middleware
+
+**Middleware** is a layer where commands and events could be intercepted before
+the moment they reach listeners. It means you can modify/cancel/extend the
+executed command in **Command Middleware**, or do appropriate operation in
+**Event Middleware** before the published event is received by its reactions.
+It's handy for logging, crash reporting, aborting particular commands or
+events, etc.
```

Comment:
From:
> **Middleware** is a layer where commands and events could be intercepted before
the moment they reach listeners. It means you can modify/cancel/extend the
executed command in **Command Middleware**, or do appropriate operation in
**Event Middleware** before the published event is received by its reactions.
It's handy for logging, crash reporting, aborting particular commands or
events, etc.

To:
> **Middleware** is a layer where commands and events can be intercepted before
they reach their listeners. It means you can modify/cancel/extend the
executed command in **Command Middleware**, or do appropriate operation in
**Event Middleware** before the published event is received by its reactions.
It's handy for logging, crash reporting, aborting particular commands or
events, etc.